bootstrap.js

```diff
@@ -4,101 +4,206 @@ Cu.import("resource://gre/modules/Console.jsm");
-      const curController = urlInputControllers.getControllerAt(i);
-      if (curController === urlInputController) {
+    const urlInput = this.browserWindow.urlInput;
+
+    for (; i < urlInput.controllers.getControllerCount(); i++) {
+      const curController = urlInput.controllers.getControllerAt(i);
+      if (curController.supportsCommand("share-button-study")) {
         i += 1;
         break;
       }
     }
-    for (i < urlInputControllers.getControllerCount(); i++;) {
-      const curController = urlInputControllers.getControllerAt(i);
+    for (; i < urlInput.controllers.getControllerCount(); i++) {
+      const curController = urlInput.controllers.getControllerAt(i);
       if (curController.supportsCommand(cmd)) {
         curController.doCommand(cmd);
         break;
       }
     }
-  },
-  onEvent(e) {},
-};
-
-function insertCopyController() {
-  // Get the "DOM" elements
-  const urlBar = bDocument.getElementById("urlbar");
-  // XUL elements are different than regular children
-  const urlInput = bDocument.getAnonymousElementByAttribute(urlBar, "anonid", "input");
+  }
 
-  // Add the controller to intercept copy "event"
-  // Store controllers so that our controller can inherit cmd_copy from
-  // the actual controller
-  urlInputControllers = urlInput.controllers;
-  urlInputControllers.insertControllerAt(0, urlInputController);
+  onEvent(e) {}
 }
 
-this.install = function(data, reason) {};
+function injectExtension(aWindow) {
+  const browserWindow = new BrowserWindow(aWindow);
+  const copyController = new CopyController(browserWindow);
 
-this.startup = function(data, reason) {
-  bWindow = Services.wm.getMostRecentWindow("navigator:browser");
-  bDocument = bWindow.document;
+  browserWindow.setCopyController(copyController);
+  browserWindowWeakMap.set(aWindow, browserWindow);
 
   // The customizationending event represents exiting the "Customize..." menu from the toolbar.
-  // We need to handle this event because after exiting the customization menu, the copy controller
-  // is removed and we can no longer detect text being copied from the URL bar.
-  // See https://dxr.mozilla.org/mozilla-central/rev/7a9536f89bc75b0672060f16ffbe6eb2c1ff3deb/browser/base/content/browser-customization.js#11
-  bWindow.addEventListener("customizationending", insertCopyController);
+  // We need to handle this event because after exiting the customization menu, the copy
+  // controller is removed and we can no longer detect text being copied from the URL bar.
+  // See DXR:browser/base/content/browser-customization.js
+  browserWindow.addCustomizeListener();
 
   // Load the CSS with the shareButton animation
-  utils = bWindow.QueryInterface(Ci.nsIInterfaceRequestor).getInterface(Ci.nsIDOMWindowUtils);
-  utils.loadSheet(CSS_URI, utils.AGENT_SHEET);
+  browserWindow.insertCSS();
 
-  insertCopyController();
+  // insert the copy controller to detect copying from URL bar
+  browserWindow.insertCopyController();
```

Comment:
After discussion, we decided to change this to `startup` and `shutdown` functions within the `BrowserWindow` class.